File: ios-native-backup/backup_20251118_215121/Nestling/Nestling/Services/UndoManager.swift
import Foundation

/// Manages undo operations for deletions.
/// Provides a lightweight undo queue with expiration.
@MainActor
class UndoManager: ObservableObject {
    static let shared = UndoManager()
    
    struct UndoableDeletion {
        let event: Event
        let deletedAt: Date
        let restoreAction: () async throws -> Void
        
        var isExpired: Bool {
            Date().timeIntervalSince(deletedAt) > 7.0 // 7 second window
        }
    }
    
    @Published private(set) var pendingDeletion: UndoableDeletion?
    
    private init() {}
    
    /// Register a deletion for potential undo.
    /// - Parameters:
    ///   - event: The event that was deleted
    ///   - restoreAction: Async closure to restore the event
    func registerDeletion(event: Event, restoreAction: @escaping () async throws -> Void) {
        pendingDeletion = UndoableDeletion(
            event: event,
            deletedAt: Date(),
            restoreAction: restoreAction
        )
        
        // Auto-expire after 7 seconds
        Task {
            try? await Task.sleep(nanoseconds: 7_000_000_000)
            if let deletion = pendingDeletion, deletion.event.id == event.id {
                await MainActor.run {
                    if pendingDeletion?.event.id == event.id {
                        pendingDeletion = nil
                    }
                }
            }
        }
    }
    
    /// Perform undo operation.
    func undo() async throws {
        guard let deletion = pendingDeletion, !deletion.isExpired else {
            throw UndoError.expired
        }
        
        try await deletion.restoreAction()
        pendingDeletion = nil
    }
    
    /// Clear pending deletion (called when user dismisses or time expires).
    func clear() {
        pendingDeletion = nil
    }
    
    /// Check if there's an undoable deletion.
    var hasUndoableDeletion: Bool {
        guard let deletion = pendingDeletion else { return false }
        return !deletion.isExpired
    }
}

enum UndoError: LocalizedError {
    case expired
    
    var errorDescription: String? {
        switch self {
        case .expired:
            return "Undo window has expired"
        }
    }
}

File: src/lib/undoManager.ts
/**
 * Undo Manager for web app
 * Provides a lightweight undo queue with 7-second expiration window
 * Matches iOS UndoManager pattern
 */

export interface UndoableDeletion<T = any> {
  item: T;
  deletedAt: Date;
  restoreAction: () => Promise<void> | void;

  isExpired(): boolean;
}

class UndoManager {
  private static instance: UndoManager;
  private pendingDeletion: UndoableDeletion | null = null;
  private expirationTimer: ReturnType<typeof setTimeout> | null = null;
  private readonly UNDO_WINDOW_MS = 7000; // 7 seconds

  private constructor() {}

  static getInstance(): UndoManager {
    if (!UndoManager.instance) {
      UndoManager.instance = new UndoManager();
    }
    return UndoManager.instance;
  }

  /**
   * Register a deletion for potential undo
   * @param item - The item that was deleted
   * @param restoreAction - Function to restore the item
   */
  registerDeletion<T>(item: T, restoreAction: () => Promise<void> | void): UndoableDeletion<T> {
    // Clear any existing pending deletion
    this.clear();

    const deletion: UndoableDeletion<T> = {
      item,
      deletedAt: new Date(),
      restoreAction,
      isExpired: () => {
        const elapsed = Date.now() - deletion.deletedAt.getTime();
        return elapsed > this.UNDO_WINDOW_MS;
      },
    };

    this.pendingDeletion = deletion;

    // Auto-expire after 7 seconds
    this.expirationTimer = setTimeout(() => {
      if (this.pendingDeletion === deletion) {
        this.clear();
      }
    }, this.UNDO_WINDOW_MS);

    return deletion;
  }

  /**
   * Perform undo operation
   * @throws Error if undo window has expired
   */
  async undo(): Promise<void> {
    if (!this.pendingDeletion) {
      throw new Error('No pending deletion to undo');
    }

    if (this.pendingDeletion.isExpired()) {
      this.clear();
      throw new Error('Undo window has expired');
    }

    try {
      await this.pendingDeletion.restoreAction();
      this.clear();
    } catch (error) {
      this.clear();
      throw error;
    }
  }

  /**
   * Clear pending deletion (called when user dismisses or time expires)
   */
  clear(): void {
    if (this.expirationTimer) {
      clearTimeout(this.expirationTimer);
      this.expirationTimer = null;
    }
    this.pendingDeletion = null;
  }

  /**
   * Check if there's an undoable deletion
   */
  hasUndoableDeletion(): boolean {
    if (!this.pendingDeletion) {
      return false;
    }
    return !this.pendingDeletion.isExpired();
  }

  /**
   * Get the pending deletion (for UI display)
   */
  getPendingDeletion(): UndoableDeletion | null {
    if (!this.pendingDeletion || this.pendingDeletion.isExpired()) {
      return null;
    }
    return this.pendingDeletion;
  }

  /**
   * Get remaining time in milliseconds
   */
  getRemainingTime(): number {
    if (!this.pendingDeletion || this.pendingDeletion.isExpired()) {
      return 0;
    }
    const elapsed = Date.now() - this.pendingDeletion.deletedAt.getTime();
    return Math.max(0, this.UNDO_WINDOW_MS - elapsed);
  }
}

export const undoManager = UndoManager.getInstance();
